---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

version <- as.vector(read.dcf('DESCRIPTION')[, 'Version'])
version <- gsub('-', '.', version)
```

# snpR

<!-- badges: start -->
[![license](https://img.shields.io/badge/license-MIT + file LICENSE-lightgrey.svg)](https://choosealicense.com/)
[![packageversion](https://img.shields.io/badge/Package%20version-`r version`-orange.svg?style=flat-square)](commits/master)
[![CRAN status](https://www.r-pkg.org/badges/version/snpR)](https://CRAN.R-project.org/package=snpR)
[![R-CMD-check](https://github.com/hemstrow/snpR/workflows/R-CMD-check/badge.svg)](https://github.com/hemstrow/snpR/actions)
<!-- badges: end -->

snpR is an R package for analyzing call Single Nucleotide Polymorphism (SNP) genotypes containing most basic stats including pairwise LD, gaussian sliding window analysis tools, plotting options, clustering analysis, colony interface, Ne estimation, formatting, filtering, and more!

## Installation

snpR can be installed from [GitHub](https://github.com/hemstrow/snpR):

``` r
# install.packages("remotes")
remotes::install_github("hemstrow/snpR")
```

To install the vignettes as well (recommended for new users), instead use:

``` r
remotes::install_github("hemstrow/snpR", build_vignettes = T) # linux
remotes::install_github("hemstrow/snpR", ref = "dev", build_vignettes = T, build_opts = c("--no-resave-data", "--no-manual")) # windows

```

The dev version can be installed from [GitHub](https://github.com/hemstrow/snpR) as well:

``` r
# install.packages("remotes")
remotes::install_github("hemstrow/snpR", ref = "dev")
```

A CRAN version should be available soon.

## Example

snpR is focused on ease-of-use. Primarily, it achieves this via the use of \emph{facets}, which describe sample or SNP metadata. snpR is built to automatically split up analysis by facet. For example, calculating genetic diversity (pi) for each population or family, or for each population/family combination is easy!


```{r example}
library(snpR)
## basic example code

x <- calc_pi(stickSNPs, facets = c("pop")) # split by pop (stickSNPs is an example dataset included in snpR)
x <- calc_pi(x, facets = c("fam")) # split by family
x <- calc_pi(x, facets = c("pop.fam")) # split by combinations of family and pop
```

snpR also facilitates ease-of-use by being \emph{overwrite safe}. As above, new analyses are added to an existing object. Results can be fetched using the get.snpR.stats handler.

```{r show}
head(get.snpR.stats(x))

```

For a full introduction, check the snpR_introduction vignette. 

```{r vignette}
# remotes::install_github("hemstrow/snpR", ref = "dev", build_vignettes = T, build_opts = c("--no-resave-data", "--no-manual"))

vignette("snpR_introduction")
```

